crates/daphne/src/messages/mod.rs

```diff
@@ -594,7 +611,86 @@ pub enum TransitionFailure {
     ReportTooEarly = 9,
 }
 
-impl TryFrom<u8> for TransitionFailure {
+/// Transition error.
+#[derive(Clone, Copy, Debug, Deserialize, Serialize, PartialEq, Eq, thiserror::Error, Hash)]
+#[serde(rename_all = "snake_case")]
+#[cfg_attr(any(test, feature = "test-utils"), derive(deepsize::DeepSizeOf))]
+pub enum TransitionFailureLatest {
+    Reserved = 0,
+    BatchCollected = 1,
+    ReportReplayed = 2,
+    ReportDropped = 3,
+    HpkeUnknownConfigId = 4,
+    HpkeDecryptError = 5,
+    VdafPrepError = 6,
+    TaskExpired = 7,
+    InvalidMessage = 8,
+    ReportTooEarly = 9,
+    TaskNotStarted = 10,
+}
+
+/// Transition error.
+#[derive(Clone, Copy, Debug, Deserialize, Serialize, PartialEq, Eq, thiserror::Error, Hash)]
+#[serde(rename_all = "snake_case")]
+#[cfg_attr(any(test, feature = "test-utils"), derive(deepsize::DeepSizeOf))]
+pub enum TransitionFailure {
+    Draft09(TransitionFailureDraft09),
+    DraftLatest(TransitionFailureLatest),
+}
+
+impl ParameterizedDecode<DapVersion> for TransitionFailure {
+    fn decode_with_param(
+        version: &DapVersion,
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        match *version {
+            DapVersion::Draft09 => match TransitionFailureDraft09::decode(bytes) {
+                Ok(x) => Ok(Self::Draft09(x)),
+                Err(x) => Err(x),
+            },
+            DapVersion::Latest => match TransitionFailureLatest::decode(bytes) {
+                Ok(x) => Ok(Self::DraftLatest(x)),
+                Err(x) => Err(x),
+            },
```

Comment:
```suggestion
            DapVersion::Draft09 => match TransitionFailureDraft09::decode(bytes).map(Self::Draft09),
            DapVersion::Latest => TransitionFailureLatest::decode(bytes).map(Self::DraftLatest),
```

This works because when you define a tuple struct (`struct Foo(i32, f32)`) rust also defines for you a function of the same name as the type (`fn Foo(a: i32, b: f32) -> Foo`), the same is true for tuple enum variants.